networkx/algorithms/tree/operations.py

```diff
@@ -7,6 +7,7 @@
 __all__ = ["join"]
 
 
+# @nx._dispatch(graphs=...)  # Not implemented
 def join(rooted_trees, label_attribute=None):
```

Comment:
perhaps an alternative version of this function that takes a single `nx.Graph` $T$ as input where $T$ is a tree with root $u$, and $u$'s children are exactly the roots of the trees in the current `join` argument, and each of the trees in the current input become subtrees under their respective roots in the new input.

Reply:
Thanks for taking a look. Yeah, having an alternative version of this (and other) functions is an option. Another alternative version could have two arguments--a list of graphs and a list of root nodes, such as `def join_trees(trees, roots): return join(list(zip(trees, roots))` (interestingly, the first step in `join` does the opposite: `trees, roots = zip(*rooted_trees)`).

If we don't want to change the NetworkX API, perhaps we need to come up with a way to say "this uses custom dispatching logic". For example:
```python
@nx._dispatch(graphs="<custom>", preserve_edge_attrs=True, preserve_node_attrs=True, preserve_graph_attrs=True)
def join(rooted_trees, label_attribute=None):
```

Reply:
Perhaps there could be a multi-hop dispatch, kind of a functional programming solution, where first the function is dispatched to an alternative networkx implementation which can be more easily mapped to the backend. This would need a new decorator `_dispatch_via`.

```python
@nx._dispatch_via(_dispatchable_f, pre_dispatch=_pre_dispatch_f, post_dispatch=_post_dispatch_f)
def hard_to_dispatch_f(*a, **kw):
   ...


def _pre_dispatch_f(*a, **kw):
   ...

@nx._dispatch(...)
def _dispatchable_f(*b, **kw):
   ...

def _post_dispatch_f(*c, **kw):
   ...

```